feat(sonarr)!: migrate series output

Fuochi opened the pull request:
Hi @davidnewhall,
These the proposed changes to series:

- using a single struct for series even if some responses and some inputs can have different empty values (mainly changes AddSeries)
- in UpdateSeries remove the ID input and leverage the ID inside the struct
- add DeleteSeries. I enforced as general behaviour to remove files and not exclude from importlist. If you prefer, we can pass the two flags as inputs.
- moving time.Time type to reference in struct. I know it's not best practice to use reference instead of values for time. However, *time.Time with "omitempty" tag is really omitted in json, while time.Time is still enforced with a 0 value. This allow us to omit time.Time fields in requests.
- add series tests
- I had to move the request checks before the response checks in GetMockServer. It seems that once the request is closed and the response is ready, the request cannot be read anymore. Still, I don't understand why it didn't trigger the error for the other tests we did.

p.s.
Sorry for the multiple builds. If you prefer, I can go back in using the forked repository.

davidnewhall commented on sonarr/series.go:
```diff
@@ -27,50 +105,52 @@ func (s *Sonarr) GetSeries(tvdbID int64) ([]*Series, error) {
-		return fmt.Errorf("api.Put(series): %w", err)
+	uri := path.Join(bpSeries, strconv.Itoa(int(series.ID)))
+	if _, err := s.PutInto(ctx, uri, params, &body, &output); err != nil {
+		return nil, fmt.Errorf("api.Put(series): %w", err)
```
A new pattern I'm going to do is this:
```suggestion
		return nil, fmt.Errorf("api.Put(%s): %w", uri, err)
```

Fuochi replied:
ok, I'll change that

davidnewhall commented on sonarr/series.go:
```diff
@@ -27,50 +105,52 @@ func (s *Sonarr) GetSeries(tvdbID int64) ([]*Series, error) {
 
 // AddSeries adds a new series to Sonarr.
-func (s *Sonarr) AddSeries(series *AddSeriesInput) (*AddSeriesOutput, error) {
+func (s *Sonarr) AddSeries(series *Series) (*Series, error) {
```
I'm not sure this is a change I want to do. Did you test this against Sonarr?

davidnewhall replied:
Yeah I looked more. These two structs are wildly different and will inevitably lead a confused user. Adding a series requires like two files, and this change makes it look like it's dozens. This may make the code cleaner, but it does not make the user experience better imho.

davidnewhall replied:
This also adds things like `AddSeriesOptions` to the output, which is just more confusion.

Fuochi replied:
To be honest, I wasn't sure about the `AddSeriesOption` as well, I'll make few more tests to be sure about it, and maybe find a better approach.
However, I don't think the two structs are much different. Actually, I think that they are mainly the same apart from `AddSeriesOption`. To create a working series you need at least 8 parameters, and many others can be configured, I still haven't checked all of them, but I'll try few things before coming back to you.
One of the biggest advantage of using the same struct is that you could take the output of a lookup and pass it directly to the AddSeries.

davidnewhall commented on shared.go:
```diff
@@ -101,7 +101,7 @@ type Tag struct {
 type Image struct {
 	CoverType string `json:"coverType"`
-	URL       string `json:"url"`
+	URL       string `json:"url,omitempty"`
```
Every image has a URL as far as I know. What does this change fix?

Fuochi replied:
In requests, it seems that RemoteURL is required while URL isn't, but I'll doublecheck this one as well.

davidnewhall replied:
oh I see. yeah some apps use URL some use RemoteURL. this is probably fine.

davidnewhall commented on sonarr/series.go:
```diff
@@ -6,11 +6,89 @@ import (
+	TotalEpisodeCount int        `json:"totalEpisodeCount"`
+	SizeOnDisk        int64      `json:"sizeOnDisk"`
+	PercentOfEpisodes float64    `json:"percentOfEpisodes"`
+	PreviousAiring    *time.Time `json:"previousAiring"`
```
This is an output struct, making time a pointer has no value here.

Fuochi replied:
right, I can change that back, in request you'll probably omit the whole field

davidnewhall commented on sonarr/series.go:
```diff
@@ -6,11 +6,89 @@ import (
+	Title             string            `json:"title,omitempty"`
+	TitleSlug         string            `json:"titleSlug,omitempty"`
+	RootFolderPath    string            `json:"rootFolderPath,omitempty"`
+	Added             *time.Time        `json:"added,omitempty"`
+	FirstAired        *time.Time        `json:"firstAired,omitempty"`
+	NextAiring        *time.Time        `json:"nextAiring,omitempty"`
+	PreviousAiring    *time.Time        `json:"previousAiring,omitempty"`
```
These are all outputs, and making them pointers has no value.

Fuochi replied:
If the struct is the same for input and output, than having a pointer will allow to omit them in the request

davidnewhall replied:
Is it necessary though? The apps may ignore these if you try to add them. 

But also, this goes with the problem above. This is _a lot of change_ to make these methods "reuse" the same data. When it makes sense to make things line up, I do it, but I've avoided overloading inputs with large outputs because it requires concessions like this, and for what actual gain?*  When you use things like AWS Go library, these "different" inputs and outputs seem to make sense.

* You mentioned being able to copy an output directly to an input to add something. I assume this means "copying from one system to another." This is a sane argument, but you're expecting end users to have some idea which fields and needed and which aren't when they go to try to change things. When they see a date like `Added` set to "something" and wonder "do I need to change that" it becomes awkward.

Fuochi replied:
I was referring to the chance to take the output of a `Lookup` method and pass it directly to the `AddSeries` method.
Something like:
```go
lookupSeries, _ := client.GetSeriesLookup("", tvdbID)
series, _ := client.AddSeries(lookupSeries[0])
```
I haven't tried it yet, but it should work.
This way, a user could be able to add a series just by knowing its tvdbID, or even just the name (if they want to make some other checks).

I'd say that this is the main advantage of having a single struct. A user can onboard even without knowing the meaning of the fields. Maybe I have been too aggressive with the `omitempty` tag, it could be used to identify which parameters are actually mandatory.

About the AWS library, I think that they are following a different approach. As far as I understand, they are defining for each api call both input and output, reusing only the nested types. 
Then, for each struct they define different methods, like validators, when needed. If you want to follow this approach, it's fine but then we should probably do that for each api call.

davidnewhall replied:
If the goal is this 
```go
lookupSeries, _ := client.GetSeriesLookup("", tvdbID) 
series, _ := client.AddSeries(lookupSeries[0])
```

You could just add another method. `AddSeriesByTvDbID(tvdbID)`

Fuochi replied:
Here I am, sorry for the late reply, busy couple of weeks.
The goal was mainly to be consistent across all resources. the easier management between methods would've been just a plus.
However, If you like to have different structs between input and output, it's fine for me. I'll try to change this PR accordingly. And I'll keep that in mind for other resources as well (even the profiles I added before should probably be changed accordingly).
Unfortunately, it will take some time, since I'll have to manually test each field if it can be a valid input or not.
I'll be back with the changes!
thanks

Fuochi replied:
Hi @davidnewhall, I made few changes at this PR to have different structs between input and output as you requested.
- I changed back the time.Time to value instead of pointer
- The input is the same for both POST and PUT, do you think we should separate the 2 of them due to AddSeriesOption?
Let me know if you prefer a different naming

davidnewhall replied:
I recently started using the `main` branch in notifiarr, and having to switch addTag and updateTag to use structs as input really sucked and didn't make any sense. Just wanted to put that out there. Thanks for updating this PR, I'll take a look shortly.

davidnewhall replied:
> However, If you like to have different structs between input and output, it's fine for me. 

This is just how the Starr apps are. I want to be consistent with _them_.

> The input is the same for both POST and PUT, do you think we should separate the 2 of them

No.

> Let me know if you prefer a different naming

I left comments about the struct names. Once you fix those I'll test again, and make sure things are still working as expected.

Fuochi replied:
Ok, Just committed the changes and rebased, let me know if you have other feedbacks

davidnewhall replied:
Would be easier to review the changes if you hadn't force-pushed. :)

Fuochi replied:
sorry, my bad! I'll try to keep in mind that I have to rebase only at the end!

davidnewhall reviewed (changes requested):
Feel free to jump on my discord and lets discuss. :)

davidnewhall commented:
I sort of feel like all the `omitempty` tags are confusing. I used those in ways they were not intended. :)

davidnewhall commented on test_methods.go:
```diff
@@ -23,20 +23,30 @@ type TestMockData struct {
 	WithError       error       // Caller's response.
 }
 
+const (
```
None of these are used?

Fuochi replied:
I used only one for now in tests, maybe it's not the best place to put them. I had in plan to use others since, 404 isn't shouldn't be returned from some calls 
https://github.com/golift/starr/blob/cfcf503e465fee35b62ea82d40062f0b55f15eab/sonarr/series_test.go#L526

davidnewhall commented:
This seems better now. I'm going to test it out in notifiarr today and see how much "changing" I need to deal with.

davidnewhall commented on sonarr/series.go:
```diff
@@ -6,71 +6,173 @@ import (
 )
 
+// Series the /api/v3/series endpoint.
+type Series struct {
```
I'd like to at least keep the struct names the same. This one should be named `AddSeriesInput`, and `SeriesOutput` should be `Series`. This will prevent library consumers from having to change piles of code.

davidnewhall replied:
Changing the struct names unnecessarily changes the interfaces to all of the Get* and Lookup* methods. These are arguably the most used pieces of this library. We should avoid changing interfaces where we don't need to.

Fuochi replied:
changed the two struct names as suggested

davidnewhall commented on sonarr/series.go:
```diff
@@ -6,71 +6,173 @@ import (
 }
 
-func (s *Sonarr) UpdateSeriesContext(ctx context.Context, seriesID int64, series *Series) error {
+func (s *Sonarr) UpdateSeriesContext(ctx context.Context, series *Series) (*SeriesOutput, error) {
```
These two Update methods are pretty much the only two that needed to change.

davidnewhall commented on sonarr/series.go:
```diff
@@ -79,61 +181,77 @@ func (s *Sonarr) AddSeriesContext(ctx context.Context, series *AddSeriesInput) (
+
+func (s *Sonarr) DeleteSeriesContext(ctx context.Context, seriesID int) error {
+	params := make(url.Values)
+	params.Add("deleteFiles", "true")
+	params.Add("addImportListExclusion", "false")
```
@austinwbest quote from OP: "I enforced as general behaviour to remove files and not exclude from importlist. If you prefer, we can pass the two flags as inputs."

What do you think about this being default behavior? Just looking for more opinions.

Fuochi replied:
We can surely set them as input! If you want we can also have another method `DeleteSeriesDefault` or something like that with default values.

davidnewhall replied:
I didn't hear from Austin (forgot to ping him in Discord today), but I do like the idea of making these inputs.
Something like:
```
func (s *Sonarr) DeleteSeriesContext(ctx context.Context, seriesID int, deleteFiles, importExclude bool) error {
```

I just know someone is going to ask for this, so we may as well provide it. Add a "comment" to explain what the inputs do please. Thanks!

davidnewhall commented on sonarr/series.go:
```diff
@@ -6,11 +6,111 @@ import (
+	AddOptions *AddSeriesOptions `json:"addOptions,omitempty"`
+}
+
+type Series struct {
```
It's a little nit-picky, but these two structs could use better comments.

davidnewhall commented on sonarr/series.go:
```diff
@@ -137,3 +237,27 @@ func (s *Sonarr) Lookup(term string) ([]*Series, error) {
+	return nil
+}
+
+// Default behaviour is to set deleteFiles to true and addImportListExclusion to false.
```
Comments should start with `// DeleteSeriesDefault  ...`

Fuochi replied:
correct, I'm going to change this before merging

davidnewhall commented:
This was the only change I had to make to Notifiarr for this: 
<img width="695" alt="Screen Shot 2022-05-20 at 11 21 25 PM" src="https://user-images.githubusercontent.com/2402929/169638800-141f27a0-24f4-4dc1-8f2f-e8bd55b1a0eb.png">

Looks good!

davidnewhall commented:
Can you stop force pushing? I'd like to keep up with your changes but you keep deleting old commits.

Fuochi commented:
> Can you stop force pushing? I'd like to keep up with your changes but you keep deleting old commits.

Just rebased against master to get the fix on travis, since I was there I also added a couple of comments you asked for, but I forgot to check the line length.
btw, rebasing and squashing before merging is best practice in TBD. Since the PR was already been approved I was following that process.

davidnewhall commented:
> rebasing and squashing

Entirely unnecessary. You can click `squash and merge` when you are finished with the PR. That's what I have always done (with this repo). I don't force push into open pull requests, ever. I've never seen anyone call that a best practice.

Fuochi commented:
> Entirely unnecessary. You can click `squash and merge` when you are finished with the PR. That's what I have always done (with this repo). I don't force push into open pull requests, ever. I've never seen anyone call that a best practice. Especially pull requests that have review comments. Force pushes tend to obliterate history.

No worries, even if it is best practice in TBD, it doesn't mean you have to adopt it. It seems you prefer to use GitHub flow, right? if that's the case I'll follow that for this repo! Sorry I haven't asked that before.

davidnewhall commented:
Please just avoid force pushing after you create a pull request. I'd like to retain the history.

davidnewhall commented:
Thank you!
